src/raftstore/store/store.rs

```diff
@@ -1277,21 +1294,29 @@ impl<T: Transport, C: PdClient> Store<T, C> {
 
         let res = peer::check_epoch(peer.region(), msg);
         if let Err(Error::StaleEpoch(msg, mut new_regions)) = res {
-            // Attach the next region which might be split from the current region. But it doesn't
-            // matter if the next region is not split from the current region. If the region meta
+            // Attach the region which might be split from the current region. But it doesn't
+            // matter if the region is not split from the current region. If the region meta
             // received by the TiKV driver is newer than the meta cached in the driver, the meta is
             // updated.
-            if let Some((_, &next_region_id)) = self.region_ranges
-                .range((Excluded(enc_end_key(peer.region())), Unbounded::<Key>))
-                .next() {
-                let next_region = self.region_peers[&next_region_id].region();
-                new_regions.push(next_region.to_owned());
+            let sibling_region_id = self.find_sibling_region(peer.region());
+            if let Some(sibling_region_id) = sibling_region_id {
+                let sibling_region = self.region_peers[&sibling_region_id].region();
+                new_regions.push(sibling_region.to_owned());
             }
             return Err(Error::StaleEpoch(msg, new_regions));
         }
         res
     }
 
+    pub fn find_sibling_region(&self, region: &metapb::Region) -> Option<u64> {
+        let start = if self.cfg.right_derive_when_split {
+            Included(enc_start_key(region))
+        } else {
+            Excluded(enc_end_key(region))
+        };
+        self.region_ranges.range((start, Unbounded::<Key>)).next().map(|(_, &region_id)| region_id)
```

Comment:
Should it be `range(Unbound::<key>, start).rev()` for `right_derive_when_split == true`?